jdbc/src/main/scala/org/apache/spark/sql/sources/jdbcExtensions.scala

```diff
@@ -167,26 +192,86 @@ object JdbcExtendedUtils extends Logging {
     (schemaName, tableName)
   }
 
-  private def getTableMetadataResultSet (table: String, conn: Connection): ResultSet = {
+  private def getTableMetadataResultSet(table: String, conn: Connection): ResultSet = {
     val (schemaName, tableName) = getTableWithSchema(table, conn)
-      // using the JDBC meta-data API
+    // using the JDBC meta-data API
     conn.getMetaData.getTables(null, schemaName, tableName, null)
   }
 
-  def tableExistsInMetaData(table: String, conn: Connection,
-      dialect: JdbcDialect): Boolean = {
+  protected lazy val getCatalystTypeMethod: Method = {
+    val m = JdbcUtils.getClass.getDeclaredMethod(
+      "org$apache$spark$sql$execution$datasources$jdbc$JdbcUtils$$getCatalystType",
+      classOf[Int], classOf[Int], classOf[Int], classOf[Boolean])
+    m.setAccessible(true)
+    m
+  }
+
+  /**
+   * Is the given JDBC type (java.sql.Types) data type signed.
+   */
+  def isSigned(typeId: Int): Boolean = {
+    typeId == Types.INTEGER || typeId == Types.FLOAT || typeId == Types.DECIMAL ||
+        typeId == Types.SMALLINT || typeId == Types.BIGINT || typeId == Types.TINYINT ||
+        typeId == Types.NUMERIC || typeId == Types.REAL || typeId == Types.DOUBLE
+  }
+
+  def getTableSchema(schemaName: String, tableName: String, conn: Connection): Seq[StructField] = {
+    val rs = conn.getMetaData.getColumns(null, schemaName, tableName, null)
+    if (rs.next()) {
+      val cols = new mutable.ArrayBuffer[StructField]()
+      do {
+        // COLUMN_NAME
+        val columnName = rs.getString(4)
+        // DATA_TYPE
+        val jdbcType = rs.getInt(5)
+        // TYPE_NAME
+        val typeName = rs.getString(6)
+        // COLUMN_SIZE
+        val size = rs.getInt(7)
+        // DECIMAL_DIGITS
+        val scale = rs.getInt(9)
+        // NULLABLE
+        val nullable = rs.getInt(11) != ResultSetMetaData.columnNoNulls
+        val metadataBuilder = new MetadataBuilder()
+            .putString("name", columnName).putLong("scale", scale)
+        val columnType = SnappyDataPoolDialect.getCatalystType(jdbcType, typeName,
+          size, metadataBuilder) match {
+          case Some(t) => t
+          case None =>
+            if (jdbcType == java.sql.Types.JAVA_OBJECT) {
+              // try to get class for the typeName else fallback to Object
+              val userClass = try {
+                Utils.classForName(typeName).asInstanceOf[Class[AnyRef]]
+              } catch {
+                case _: Throwable => classOf[AnyRef]
+              }
+              new JavaObjectType(userClass)
+            } else {
+              getCatalystTypeMethod.invoke(JdbcUtils, Int.box(jdbcType),
+                Int.box(size), Int.box(scale),
+                Boolean.box(isSigned(jdbcType))).asInstanceOf[DataType]
+            }
+        }
+        cols += StructField(columnName, columnType, nullable, metadataBuilder.build())
+      } while (rs.next())
+      cols
+    } else Nil
```

Comment:
Hmm.. I don't think we can fix the problem of schema changing in the embedded data source while the spark side has built the relation using a previous definition. Wish if that was detected too.

Reply:
Isn't there the way to throw back exception if table has been altered to clear the cached relations? It should be using the local version field on embedded side and compare against what smart connector knows about. Will be a problem in thin connector though.

Reply:
Yeah, was talking about the thin connector. Or the jdbc datasource access too. Because first time spark side will fire the  metadata api to find out the schema and then it will cache the relation I guess.

Reply:
I don't think its a problem with jdbc datasource access in most cases at least the way we are proposing to use with query pushdown. Unless user fires the exact same query multiple times, else new query will lead to creation of a new relation.